Add shipment information box on fulfillment editor

tpaksu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds the shipping information entry and display boxes, and their tests to the fulfillment drawer. 

- The user can add this information while creating the fulfillment, or skip it (choose no information provided) on the creation, and add it with updating the fulfillment. 
- There are some shipping providers implemented as JS object for demo purposes, normally it will be received via a global variable like `wcSettings` from the PHP side. This will be handled on a separate issue alongside the currency symbols, the fulfillment status strings, the meta key strings, etc.
- I also added a demo lookup response, which shows up when you enter "12345678" to the tracking number look up form, this will be deleted before releasing the feature.

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes WOOPLUG-4000

### Screenshots or screen recordings:

<img width="436" alt="image" src="https://github.com/user-attachments/assets/f6dda522-1784-4f04-a828-764669ab682b" />
<img width="431" alt="image" src="https://github.com/user-attachments/assets/a3e07064-9ebd-4b6d-ad4d-23f4e12d3287" />
<img width="436" alt="image" src="https://github.com/user-attachments/assets/2a41b9c5-5d7d-41e5-95c6-563f476b8e9c" />
<img width="438" alt="image" src="https://github.com/user-attachments/assets/d6d19c8a-a6f9-46e3-b2f3-d44ee560a268" />
<img width="429" alt="image" src="https://github.com/user-attachments/assets/2b534eae-97de-40e1-8663-00cbbaaf5816" />
<img width="434" alt="image" src="https://github.com/user-attachments/assets/6f127570-4557-46c6-a75b-519fc3be6b5a" />


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

- Checkout this branch
- Go to `plugins/woocommerce` folder, and run `pnpm run watch:build` to build the assets.
- Create some orders if you don't have any in your test environment.
- Open WooCommerce > Orders page.

#### Tracking number lookup with a new fulfillment

- Test the tracking number lookup with no tracking number entered. You should see an error.
- Test the tracking number lookup with a random tracking number. You should see the same error.
- Test the tracking number lookup with "12345678", you should see a shipment provider, and a tracking link returned. Also the tracking number input will convert to a label with an edit icon.
- Note the shipment provider and tracking URL and save the fulfillment. Open the saved fulfillment, and you should see the same information displayed.
- Edit the fulfillment, and you should see the same information on the shipping information edit form.

#### Manual entry with a new fulfillment

> [!Note] 
> Those fields are not validated, and we allow any field to be empty on save. The user will have the control over that, and it's their responsibility to have them filled. We don't know the possible scenarios the user might have.

- Test the manual entry with a tracking number, a shipment provider from the list, and a tracking link. 
- Store the fulfillment and see if they are saved correctly.
- Edit the fulfillment, and see if the form opens with manual entry option selected, and the inputs have correct values.
- Test it again with selecting "Other" at the bottom of the shipment provider list. There should be a new input visible allowing you to enter a name for the shipping provider.
- Fill it with a provider that isn't available on the list, fill other fields and save.
- Check if the fulfillment displays the correct values
- Check if editing the fulfillment shows the correct values

#### No information entered

- Test saving and editing a new fulfillment with "No information" radio selected. In all steps, it should show the correct values.

#### Testing edits

For all scenarios above, create a new fulfillment with "No information provided", selected, and test the scenario applied with editing the fulfillment. It should show the correct values after editing and saving the fulfillment. 

#### Check design

Open this 2xydCrrB4yb2TAqpXt607y-fi-3939_19728 and check if the designs match.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

Adds a shipment information editor for the fulfillments.

</details>

tpaksu committed:
Initial commit

tpaksu committed:
Merge branch 'feature/57353-order-fulfillments-entity' into add/57363-order-fulfillments-base-user-interface-2

tpaksu committed:
Add fulfillments drawer renderer and component

Commit:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

tpaksu committed:
Fix comments and method naming

tpaksu committed:
Add tests

tpaksu committed:
Add check for null screen

tpaksu committed:
Test returning internal meta keyse to fix unit test error on Github

tpaksu committed:
Fix non-static method called as static method

tpaksu committed:
Remove fix attempt as it didn't work

tpaksu committed:
Fix test error caused by container mock

tpaksu committed:
Initial commit

tpaksu committed:
Add item listing and shipment information blocks

tpaksu committed:
Add checkbox selection details

tpaksu committed:
Fix bulk checkbox, add select all link

tpaksu committed:
Finish shipment information form

tpaksu committed:
Add notification box and buttons

tpaksu committed:
Add ShipmentFormContext, restructure components

tpaksu committed:
Refactor code, create fulfillments list

tpaksu committed:
Add fulfillment editor

tpaksu committed:
Fix items reverting after click, fix radio group on shipping options, keep order items visible when a fulfillment details are open

tpaksu committed:
Address multiple issues

tpaksu committed:
Merge branch 'add/57363-order-fulfillments-base-user-interface-2' into add/WOOPLUG-3977-order-fulfillments-fulfillment-editor-form

tpaksu committed:
Merge branch 'feature/57353-order-fulfillments-entity' into add/WOOPLUG-3977-order-fulfillments-fulfillment-editor-form

tpaksu committed:
Add update, delete actions, shipping form manual edit names, componentize recurring code

tpaksu committed:
Add JS tests

tpaksu committed:
Merge branch 'feature/57353-order-fulfillments-entity' into add/WOOPLUG-3977-order-fulfillments-fulfillment-editor-form

tpaksu committed:
Fix issues with icons and images, add max-width to drawer

tpaksu committed:
Split PR into multiple PR's that solve only the focused issue

tpaksu committed:
Add disabled state when editing a fulfillment

tpaksu committed:
Disable other fulfillments when editing, add disabled states to buttons, fix test, fix fulfillment UI update after edit

tpaksu committed:
Add collapse to pending items, css styling

tpaksu committed:
Fix test

tpaksu committed:
Set order items open when no fulfillments are available

Commit:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

tpaksu committed:
Add shipment information box and tests

tpaksu committed:
Fix fulfillment shipping settings not updating as expected, add constants for meta keys, disable changing values on lookup form, fix tests

Commit:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

ismaeldcom commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/shipment-form/shipment-manual-entry-form.tsx:
```diff
@@ -0,0 +1,139 @@
+			</div>
+			{ shipmentProvider === 'other' && (
+				<div className="woocommerce-fulfillment-input-container">
+					<h4>{ __( 'Name', 'woocommerce' ) }</h4>
```
Why using `h4` instead of the `label` property from `TextControl`?

tpaksu replied:
Good catch, addressed in [71d7b63](https://github.com/woocommerce/woocommerce/pull/57989/commits/71d7b6384a44cf3d74227aa498a388d288df0bcf).

ismaeldcom commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/utils/icons.tsx:
```diff
@@ -0,0 +1,108 @@
+	</svg>
+);
+
+export const CopyIcon = ( { copyText }: { copyText: string } ) => {
```
We should consider using a button here to ensure accessibility.

tpaksu replied:
Addressed in [71d7b63](https://github.com/woocommerce/woocommerce/pull/57989/commits/71d7b6384a44cf3d74227aa498a388d288df0bcf).

ismaeldcom commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillments-card/card.tsx:
Did you evaluate using the existing [Panel component](https://wordpress.github.io/gutenberg/?path=/docs/components-panel--docs)?

tpaksu replied:
Yep, I've tried that before, just tried that yesterday again to see what the issue was, and confirmed that the panel component doesn't support disabled state (need to disable other tabs when editing a fulfillment), and custom components on the title (we need to display the fulfillment status badge on the title). Once that features are added, we can replace those with the existing panel components.

ismaeldcom replied:
Thanks for checking it!

ismaeldcom commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/context/shipment-form-context.tsx:
```diff
@@ -0,0 +1,123 @@
+		defaultContextProps.providerName
+	);
+
+	// fulfillment değiştiğinde state'leri güncelle
```
This needs translation for those who don't understand Turkish. 🙂

tpaksu replied:
🤦‍♂️ Good catch 😬

tpaksu replied:
Fixed in [71d7b63](https://github.com/woocommerce/woocommerce/pull/57989/commits/71d7b6384a44cf3d74227aa498a388d288df0bcf).

ismaeldcom commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/context/shipment-form-context.tsx:
```diff
@@ -0,0 +1,123 @@
+
+	return (
+		<ShipmentFormContextValue.Provider
+			value={ {
```
Both this and the fulfillment context will benefit from `useMemo` here. Maybe is not an issue right now, but any component consuming this provider will rerender every time any these properties change even if they are not using them.

tpaksu replied:
Addressed in [71d7b63](https://github.com/woocommerce/woocommerce/pull/57989/commits/71d7b6384a44cf3d74227aa498a388d288df0bcf).

ismaeldcom commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/utils/icons.tsx:
```diff
@@ -0,0 +1,108 @@
+	>
+		<path
+			d="M11.8333 2.83301L9.33329 0.333008L2.24996 7.41634L1.41663 10.7497L4.74996 9.91634L11.8333 2.83301ZM5.99996 12.4163H0.166626V13.6663H5.99996V12.4163Z"
+			fill="#3858E9"
```
Remember to remove the `fill` here if you need to change it's color from CSS.

tpaksu replied:
I'll leave this to later design review and change if needed.

ismaeldcom reviewed (commented):
Thanks for splitting the PR to make it easier to review. The testing instructions work as intented but I've noticed some issues, and although I'm not sure if all of them fall into this PR I'll share them here.

#### UX issues
- In an order with a product x2, create a fulfillment for just one of them. Editing the fullfilment later will select the two instead of just one.
- After opening the fulfillment modal for a second order, the products' checkbox doesn't work, I need to first click on `Select all` to be able to use them again.
- If you switch to a different order fulfillment while editing one, you end up with the sidebar in edit state (header disabled)

#### Design review
Not sure if you prefer to ask for a review from Design, but here is what I've noticed by comparing the current implementation with the Figma one.

- There are no items selected by default.
- The `Select all(n)` button should display `N Selected` instead, in any case.
- After clicking on **Find info** in **Tracking Number**, provider should display the icon and name, and tracking URL should by an anchor. Right now these are disabled inputs.
- The edit (pen) button should be grey instead of blue.
- Fulfillment status should be capitalized.
- While editing a fulfillment, other fulfillments should be disabled but not the header.

It's not clear to me from the designs, but I think we should unwrap by default a fulfillment if there are no products left. Otherwise you only see a collapsed **Fulfillment 1** on the sidebar.

tpaksu committed:
Styling fixes, item price fixes

tpaksu committed:
Fix item selection

tpaksu committed:
Merge branch 'add/WOOPLUG-3977-order-fulfillments-fulfillment-editor-form' into add/57405-order-fulfillments-create-shipping-information-entry-form

tpaksu committed:
Replace H4's with component labels, apply design to tracking number lookup result,  add useMemo to context values

tpaksu committed:
Show new fulfillment form items as checked initially

tpaksu committed:
Merge branch 'add/WOOPLUG-3977-order-fulfillments-fulfillment-editor-form' into add/57405-order-fulfillments-create-shipping-information-entry-form

tpaksu commented:
> In an order with a product x2, create a fulfillment for just one of them. Editing the fullfilment later will select the two instead of just one.

Thanks for noticing, fixed it on the parent PR. 

> After opening the fulfillment modal for a second order, the products' checkbox doesn't work, I need to first click on Select all to be able to use them again.

Yep, that was also an issue which surfaced on the parent PR, fixed it there.

> If you switch to a different order fulfillment while editing one, you end up with the sidebar in edit state (header disabled)

I'm not sure how to fix this one, maybe I should check if the _editMode_ is active and prevent the order ID change, but the editMode is in the inner context. We should do a brainstorm on this.

> Not sure if you prefer to ask for a review from Design, but here is what I've noticed by comparing the current implementation with the Figma one.

I'll ask for a design review once all the components and flow are completed, and create another PR for fixing all the issues. 

> There are no items selected by default.

Thanks for noticing. I've rewritten the logic for the items tree, which should fix it in the parent PR. 

> The Select all(n) button should display N Selected instead, in any case.

Can you check the latest update if it works as you mentioned?

> After clicking on Find info in Tracking Number, provider should display the icon and name, and tracking URL should by an anchor. Right now these are disabled inputs.
> The edit (pen) button should be grey instead of blue.

Yep, fixed them in this PR. 

> Fulfillment status should be capitalized.

This is waiting for getting the translated string representations from the backend, these shown are only keys for now.

> While editing a fulfillment, other fulfillments should be disabled but not the header.

Fixed it. 

> It's not clear to me from the designs, but I think we should unwrap by default a fulfillment if there are no products left. Otherwise you only see a collapsed Fulfillment 1 on the sidebar.

What do you mean by "unwrap"? Showing all the fulfillments as expanded?

tpaksu committed:
Change icon color, undisable header, disable edit mode before closing drawer

tpaksu committed:
Skip disabling header on edit, exit edit mode on drawer close

tpaksu committed:
Merge branch 'add/WOOPLUG-3977-order-fulfillments-fulfillment-editor-form' into add/57405-order-fulfillments-create-shipping-information-entry-form

tpaksu committed:
Small fixes

tpaksu committed:
Merge branch 'add/WOOPLUG-3977-order-fulfillments-fulfillment-editor-form' into add/57405-order-fulfillments-create-shipping-information-entry-form

tpaksu committed:
Increase meta list font size

ismaeldcom reviewed (approved):
Thanks for taking care of every comment! Everything is fixed and works as expected. I noticed a small issue, but it's not related to this PR; more on that below.

>I'm not sure how to fix this one, maybe I should check if the editMode is active and prevent the order ID change, but the editMode is in the inner context. We should do a brainstorm on this.

The easiest way to fix it is adding a `key` to [the provider here](https://github.com/woocommerce/woocommerce/blob/0ab1940d042bfe47cddaac1935e2df6061c9d030/plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer.tsx#L36). Using `orderId` as the key will create a different provider per order, instead of reusing the same one and preserving its state.

While looking at `FulfillmentDrawerProvider`, I've noticed that you're using `setState` within `useSelect` while having these states in the dependencies, which causes an infinite loop that you prevented using `isEqual`.

We should return the store date, like this:
```tsx
const { order, fulfillments} = useSelect(
  ( select ) => {
    const store = select( FulfillmentsStore );
    return {
      order: store.getOrder( orderId ),
      fulfillments: store.readFulfillments( orderId ) ?? [],
    };
  },
  [ orderId ]
);
```
If you move the `orderId` check outside of the provider, you can greatly simplify it. If I'm not wrong, order and fulfillment states were only used within the provider, so they don't need to be exposed and can be removed.

tpaksu committed:
Merge branch 'feature/57353-order-fulfillments-entity' into add/57405-order-fulfillments-create-shipping-information-entry-form
